### PR TITLE
gitter.py: Create GitterRoomThread class and use it

### PR DESCRIPTION
**Pull request by meetmangukiya**
Fixes https://github.com/errbotio/err-backend-gitter/issues/23
Fixes https://github.com/errbotio/err-backend-gitter/issues/9
Fixes https://github.com/errbotio/err-backend-gitter/issues/6

### CODE REVIEW

**Comment by meetmangukiya** [comment]
@zoni this is an interesting PR :wink: I think this'll definitely fix the problem :D

**Comment by meetmangukiya** [comment]
this might not be working, completely experimental

**Comment by meetmangukiya** [comment]
Some small problems, but fixed now, works for me :+1:

File: gitter.py [hunk 1]
```diff
@@ -216,6 +224,40 @@ def __unicode__(self):
...
+                        m.to = self.backend.bot_identifier
+                    else:
+                        m = Message(json_message['text'])
+                        m.to = self.room
+                    m.frm = GitterRoomOccupant.build_from_json(self.room, from_user)
+                    self.backend.callback_message(m)
+                else:
+                    log.debug('Received keep-alive on %s', self.room.name)
+        except:
+            return
```
Comment (gbin) [inline gitter.py]: Log the exception here otherwise it might not work and the user will never know why

File: gitter.py [hunk 2]
```diff
@@ -216,6 +224,40 @@ def __unicode__(self):
...
+            for line in r.iter_lines(chunk_size=1):  # it fails with anything else than 1.
+                if line.strip():
+                    json_message = json.loads(line.decode('utf-8'))
+                    from_user = json_message['fromUser']
+                    log.debug("Raw message from room %s: %s" % (self.room.name, json_message))
+                    if self.room._uri == from_user['url']:
+                        m = Message(json_message['text'])
+                        m.to = self.backend.bot_identifier
+                    else:
+                        m = Message(json_message['text'])
```
Comment (gbin) [inline gitter.py]: Build this before the if

File: gitter.py [hunk 3]
```diff
@@ -216,6 +224,40 @@ def __unicode__(self):
...
+    def __init__(self, room, backend):
+        super().__init__()
+        self.room = room
+        self.backend = backend
+
+    def run(self):
+        self.room.joined = True
+        log.debug("thread for %s started" % self.room.idd)
+        while True:
+            self.stream()
```
Comment (gbin) [inline gitter.py]: Can we make it quit properly?

Comment (meetmangukiya) [inline gitter.py]: except `KeyboardInterrupt` ?

Comment (meetmangukiya) [inline gitter.py]: but that is done already, in `serve_once`

**Comment by meetmangukiya** [comment]
@gbin ready for next review, thanks for reviewing :smile:

File: gitter.py (hunk 3 above)
Comment (zoni) [inline gitter.py]: Don't think that's enough/applicable here.

Also, this needs incremental backoff I think. With the current loop, if `streamAPIRequest` fails immediately, we would end up hammering Gitter with multiple requests/s.

File: gitter.py [hunk 4]
```diff
@@ -179,7 +180,14 @@ def idd(self):
...
         return self._name
 
-    joined = True  # TODO
+    @property
+    def joined(self):
+        return self._joined
+
+    @joined.setter
+    def joined(self, value):
+        self._joined = value
```
Comment (zoni) [inline gitter.py]: I don't think this setter should be here. We only want outside code to get the property, not set it. It's only ever set internally from within this class.

**Pull request merged**